File: typescript/src/agents/lexBotAgent.ts
import { Agent, AgentOptions } from "./agent";
import { ConversationMessage, ParticipantRole } from "../types";
import {
  LexRuntimeV2Client,
  RecognizeTextCommand,
  RecognizeTextCommandOutput,
} from "@aws-sdk/client-lex-runtime-v2";
import { Logger } from "../utils/logger";

/**
 * Options for configuring an Amazon Lex Bot agent.
 * Extends base AgentOptions with specific parameters required for Amazon Lex.
 */
export interface LexBotAgentOptions extends AgentOptions {
  region?: string;
  botId: string; // The ID of the Lex Bot
  botAliasId: string; // The alias ID of the Lex Bot
  localeId: string; // The locale of the bot (e.g., en_US)
}

/**
 * LexBotAgent class for interacting with Amazon Lex Bot.
 * Extends the base Agent class.
 */
export class LexBotAgent extends Agent {
  private readonly lexClient: LexRuntimeV2Client;
  private readonly botId: string;
  private readonly botAliasId: string;
  private readonly localeId: string;

  /**
   * Constructor for LexBotAgent.
   * @param options - Configuration options for the Lex Bot agent
   */
  constructor(options: LexBotAgentOptions) {
    super(options);
    this.lexClient = new LexRuntimeV2Client({ region: options.region });
    this.botId = options.botId;
    this.botAliasId = options.botAliasId;
    this.localeId = options.localeId;

    // Validate required fields
    if (!this.botId || !this.botAliasId || !this.localeId) {
      throw new Error("botId, botAliasId, and localeId are required for LexBotAgent");
    }
  }

  /**
   * Process a request to the Lex Bot.
   * @param inputText - The user's input text
   * @param userId - The ID of the user
   * @param sessionId - The ID of the current session
   * @param chatHistory - The history of the conversation
   * @param additionalParams - Any additional parameters to include
   * @returns A Promise resolving to a ConversationMessage containing the bot's response
   */
  /* eslint-disable @typescript-eslint/no-unused-vars */
  async processRequest(
    inputText: string,
    userId: string,
    sessionId: string,
    chatHistory: ConversationMessage[],
    additionalParams?: Record<string, string>
  ): Promise<ConversationMessage> {
    try {
      // Prepare the parameters for the Lex Bot request
      const params = {
        botId: this.botId,
        botAliasId: this.botAliasId,
        localeId: this.localeId,
        sessionId: sessionId,
        text: inputText,
        sessionState: {
          // You might want to maintain session state if needed
        },
      };

      // Create and send the command to the Lex Bot
      const command = new RecognizeTextCommand(params);
      const response: RecognizeTextCommandOutput = await this.lexClient.send(command);

      // Process the messages returned by Lex
      let concatenatedContent = '';
      if (response.messages && response.messages.length > 0) {
        concatenatedContent = response.messages
          .map(message => message.content)
          .filter(Boolean)
          .join(' ');
      }

      // Construct and return the Message object
      return {
        role: ParticipantRole.ASSISTANT,
        content: [{ text: concatenatedContent || "No response from Lex bot." }],
      };
    } catch (error) {
      // Log the error and re-throw it
      Logger.logger.error("Error processing request:", error);
      throw error;
    }
  }
}

File: python/src/multi_agent_orchestrator/agents/lex_bot_agent.py
from typing import List, Dict, Optional
from dataclasses import dataclass
import boto3
from botocore.exceptions import BotoCoreError, ClientError
from multi_agent_orchestrator.agents import Agent, AgentOptions
from multi_agent_orchestrator.types import ConversationMessage, ParticipantRole
from multi_agent_orchestrator.utils import Logger
import os

@dataclass
class LexBotAgentOptions(AgentOptions):
    bot_id: str = None
    bot_alias_id: str = None
    locale_id: str = None

class LexBotAgent(Agent):
    def __init__(self, options: LexBotAgentOptions):
        super().__init__(options)
        if (options.region is None):
            self.region = os.environ.get("AWS_REGION", 'us-east-1')
        else:
            self.region = options.region
        self.lex_client = boto3.client('lexv2-runtime', region_name=self.region)
        self.bot_id = options.bot_id
        self.bot_alias_id = options.bot_alias_id
        self.locale_id = options.locale_id

        if not all([self.bot_id, self.bot_alias_id, self.locale_id]):
            raise ValueError("bot_id, bot_alias_id, and locale_id are required for LexBotAgent")

    async def process_request(self, input_text: str, user_id: str, session_id: str,
                        chat_history: List[ConversationMessage],
                        additional_params: Optional[Dict[str, str]] = None) -> ConversationMessage:
        try:
            params = {
                'botId': self.bot_id,
                'botAliasId': self.bot_alias_id,
                'localeId': self.locale_id,
                'sessionId': session_id,
                'text': input_text,
                'sessionState': {}  # You might want to maintain session state if needed
            }

            response = self.lex_client.recognize_text(**params)

            concatenated_content = ' '.join(
                message.get('content', '') for message in response.get('messages', [])
                if message.get('content')
            )

            return ConversationMessage(
                role=ParticipantRole.ASSISTANT.value,
                content=[{"text": concatenated_content or "No response from Lex bot."}]
            )

        except (BotoCoreError, ClientError) as error:
            Logger.error(f"Error processing request: {str(error)}")
            raise error

